pyquil/quilbase.py

```diff
@@ -913,167 +919,283 @@ class ClassicalMove(AbstractInstruction):
+
+    @right.setter
+    def right(self, right: Union[MemoryReference, int, float]) -> None:
+        quil_rs.Store.source.__set__(self, _to_rs_arithmetic_operand(right))  # type: ignore
 
     def out(self) -> str:
-        return "%s %s %s %s" % (self.op, self.target, self.left, self.right)
+        return str(self)
 
 
-class ClassicalComparison(AbstractInstruction):
+class ClassicalComparison(quil_rs.Comparison, AbstractInstruction):
     """
     Abstract class for ternary comparison instructions.
     """
 
-    op: ClassVar[str]
+    op: ClassVar[quil_rs.ComparisonOperator]
 
-    def __init__(
-        self,
+    def __new__(
+        cls,
         target: MemoryReference,
         left: MemoryReference,
         right: Union[MemoryReference, int, float],
-    ):
-        if not isinstance(target, MemoryReference):
-            raise TypeError("target operand should be an MemoryReference")
-        if not isinstance(left, MemoryReference):
-            raise TypeError("left operand should be an MemoryReference")
-        if not (isinstance(right, MemoryReference) or isinstance(right, int) or isinstance(right, float)):
-            raise TypeError("right operand should be an MemoryReference or an int or float.")
-        self.target = target
-        self.left = left
-        self.right = right
+    ) -> "ClassicalComparison":
+        operands = (target._to_rs_memory_reference(), left._to_rs_memory_reference(), cls._to_comparison_operand(right))
+        return super().__new__(cls, cls.op, operands)
+
+    @staticmethod
+    def _to_comparison_operand(operand: Union[MemoryReference, int, float]) -> quil_rs.ComparisonOperand:
+        if isinstance(operand, MemoryReference):
+            return quil_rs.ComparisonOperand.from_memory_reference(operand._to_rs_memory_reference())
+        elif isinstance(operand, int):
+            return quil_rs.ComparisonOperand.from_literal_integer(operand)
+        elif isinstance(operand, float):
+            return quil_rs.ComparisonOperand.from_literal_real(operand)
+        raise TypeError(f"{type(operand)} is not a valid ComparisonOperand")
+
+    @staticmethod
+    def _to_py_operand(operand: quil_rs.ComparisonOperand) -> Union[MemoryReference, int, float]:
+        if not isinstance(operand, quil_rs.ComparisonOperand):
+            raise TypeError(f"{type(operand)} is not an ComparisonOperand")
+        inner = operand.inner()
+        if isinstance(inner, quil_rs.MemoryReference):
+            return MemoryReference._from_rs_memory_reference(inner)
+        return inner
+
+    @property
+    def target(self) -> MemoryReference:
+        return MemoryReference._from_rs_memory_reference(super().operands[0])
+
+    @target.setter
+    def target(self, target: MemoryReference) -> None:
+        operands = list(super().operands)
+        operands[0] = target._to_rs_memory_reference()
+        quil_rs.Comparison.operands.__set__(self, tuple(operands))  # type: ignore
```

Comment:
Why not `tuple` operands from the beginning?

Reply:
Tuples are immutable, so `operands[0] = …` wouldn't work.